javalanglib/src/main/scala/java/lang/Float.scala

```diff
@@ -72,8 +76,229 @@ object Float {
+   *  because we can leverage Double arithmetics to parse only a Float. In
+   *  particular, we can accurately compute and represent the two adjacent
+   *  Floats that enclose the best approximation, as well as the midpoint
+   *  between those, which is a Double. We receive those from
+   *  `parseFloatDecimal`, which already had to compute them in order to decide
+   *  whether a correction was needed. The only real thing we keep from the
+   *  paper is the step 3: how to accurately compare that midpoint with the
+   *  exact value represented by the string, using big integer arithmetics.
+   *  This allows us to decide whether we need to round up, down, or break a
+   *  tie to even.
+   *
+   *  `AlgorithmR` in the paper is generic wrt. the bases of the input and
+   *  output. In our case, the input base Δ is 10 and the output base β is 2.
+   */
+  private def parseFloatDecimalCorrection(integralPartStr: String,
+      fractionalPartStr: String, exponentStr: String,
+      zDown: scala.Float, zUp: scala.Float, mid: scala.Double): scala.Float = {
+
+    // 1. Accurately parse the string with the representation f × 10ᵉ
+
+    val f: BigInteger = new BigInteger(integralPartStr + fractionalPartStr)
+    val e: Int = Integer.parseInt(exponentStr) - fractionalPartStr.length()
+
+    /* Note: we know that `e` is "reasonable". If it were way too big or way
+     * too small, the original JS `parseFloat` in `parseFloatDecimal` would
+     * have returned `Infinity` or `0.0`, respectively. In that case, we would
+     * have selected the first branch, and never called
+     * `parseFloatDecimalCorrection`.
+     */
+
+    // 2. Accurately decompose `mid` with the representation m × 2ᵏ
+
+    val mbits = 52 // number of bits of the mantissa (without the implicit '1')
+    val kbits = 11 // number of bits of the exponent
+    val bias = (1 << (kbits - 1)) - 1 // the bias of the exponent
+
+    val midBits = Double.doubleToLongBits(mid)
+    val biasedK = (midBits >> mbits).toInt
+
+    /* Because `mid` is a double value halfway between two floats, it cannot
+     * be a double subnormal (even if the two floats that surround it are
+     * subnormal floats).
+     */
+    if (biasedK == 0)
+      throw new AssertionError("parseFloatCorrection was given a subnormal mid: " + mid)
+
+    val m = BigInteger.valueOf((midBits & ((1L << mbits) - 1)) | (1L << mbits))
+    val k = biasedK - bias - mbits
+
+    // 3. Accurately compare f × 10ᵉ to m × 2ᵏ
+
+    @inline def compare(x: BigInteger, y: BigInteger): Int =
+      x.compareTo(y)
+
+    val cmp = if (e >= 0) {
+      if (k >= 0)
+        compare(multiplyBy10Pow(f, e), multiplyBy2Pow(m, k))
+      else
+        compare(multiplyBy2Pow(multiplyBy10Pow(f, e), -k), m) // this branch may be dead code in practice
+    } else {
+      if (k >= 0)
+        compare(f, multiplyBy2Pow(multiplyBy10Pow(m, -e), k))
+      else
+        compare(multiplyBy2Pow(f, -k), multiplyBy10Pow(m, -e))
+    }
```

Comment:
IIUC, this is essentially home-brewn BigDecimal logic. Do we have a grasp on how much more expensive (in code size I assume) it would be to just actually use `BigDecimal`s for this?

Reply:
I don't think `BigDecimal` is up to the job, here. `BigDecimal`s are represented with a base 10 scale, so they can accurately and efficiently represent f × 10ᵉ, but they do not have a native way of representing m × 2ᵏ. Just [the code to convert a `Double` into `BigDecimal`](https://github.com/scala-js/scala-js/blob/4e50422bf8013cc1995327abba245cf3dc27f5b5/javalib/src/main/scala/java/math/BigDecimal.scala#L470-L519) is way more complicated than the code we have here.

Also, looking at the future, I think there's a qualitative difference between `BigInteger` and `BigDecimal`. Since recent JS has `bigint`s, we will eventually have ESFeatures that allow Scala.js to rely on `bigint`s. In that case, this code would have 0 code size overhead, whereas code relying on `BigDecimal`s would.

Reply:
Right. BigDecimal is only arbitrary precision for decimal representation, not for binary.